### DISCUSSION TITLE
handling Map<String,Object> of known types but not polymorphic

### DISCUSSION THREAD

**Discussion by robshep** [FasterXML/jackson-databind#4330] [General]
Hi,

I have a Map<String,Object> property in a few classes.

I'm trying to workout how to deserialise this effectively, but cannot find a hook to do, pointers appreicated.

My values are not polymorphic, but they could be primitives (double, int, long, boolean, Sring) or a small set of unrelated objects.

In one instance it's OK for me to add type info.  I've used `defaultTyping(...)` with a `TypeResolverBuilder` and it serialises type info into the  values where needed.  This deserialises properly too.

But this approach has an effect on any generic object, and so other instances of Map<String,Object> fail to deserialise.

It works if I add type info into the JSON as prescribed, but I can't alter some clients who use these interfaces....  

(Note: it's mostly things like configurations where most of the time its a primitive, but occasionally some structured data that will evolve slowly)

I could use a different customised ObjectMapper in a few instances if necessary. 

Otherwise, is there a (mostly) de-serialisation hook I can use on something like that 

```java
@JsonXYZ(MyMapObjectValueDeserialiser.class)
private Map<String,Object> config;
```

I'll also know how to deserialise these values by key name - we have key name conventions for these different types.

Thanks in advance for nay pointers.

R

**Comment by cowtowncoder** [comment]
Just to make sure: have you tried adding `@JsonTypeInfo` directly on Map-valued property? That should only enable Polymorphic handling for values of that `Map`?

Alternatively if you wanted to go custom deserializer route, you can similarly annotated one used for Map values (but not Map itself) with

```
@JsonDeserialize(contentUsing = MyObjectValueDeserializer.class)
Map<String, Object> stuff;
```

One challenge in both cases, though, is that value deserializer has no access to key of `Map`: for that you would need custom deserializer for Map itself (that is, `@JsonDeserialize(using = MyMapDeserializer.class)`).

**Comment by robshep** [comment]
Thank you @cowtowncoder 

> have you tried adding @JsonTypeInfo directly on Map-valued property?

Wow, I thought I'd tried every annotation - everywhere....  

but seeingly not this combination.

Thank you

```java
@Test
public void testJacksonTypeInfo() throws JsonProcessingException {
    ThingWithMap twm1 = new ThingWithMap();
    twm1.setMyMap1(Map.of("mk1", new MapValue1(), "mk2", new MapValue2(), "i1", 42, "s1", "a string", "b1", true));
    twm1.setMyMap2(Map.of("mk1", new MapValue1(), "mk2", new MapValue2(), "i1", 42, "s1", "a string", "b1", true));
    String jsonString = objectMapper.writeValueAsString(twm1);
    Assertions.assertThat(jsonString).containsOnlyOnce("\"_class\":\"lang.Jackson$MapValue1\"");
    Assertions.assertThat(jsonString).containsOnlyOnce("\"_class\":\"lang.Jackson$MapValue2\"");
}

@Data
public static class ThingWithMap
{
    @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "_class")
    private Map<String,Object> myMap1 = new HashMap<>();

    private Map<String,Object> myMap2 = new HashMap<>();
}

@Data
public static class MapValue1
{
    private String name = "rob";
    private int age = -1;
}

@Data
public static class MapValue2
{
    private Boolean enabled = true;
    private int volume = 11;
}

```

**Reply by cowtowncoder** [reply]
Ok so what happens in that case? Another note is that it looks like Lombok is being used -- sometimes that affects results so if possible it'd be good to try without `@Data` annotation (adding necessary accessors manually) and see if there is any difference.
Lombok has many options for copying annotations during bytecode processing, I think, but otherwise it is invisible to Jackson.

**Reply by robshep** [reply]
Thank you,
The tests above provide exactly what I need.

`myMap1` has the `@JsonTypeInfo` annotation configuration, 
`myMap2` does not.  

The output is excellent with embedded type info on map values within the annotated map property `myMap1`, 
but no embedded type info on the map values within the non-annotated map property `myMap2`. 

I can extend the test case to show the reverse:

```java
ThingWithMap twnRx = objectMapper.readValue(jsonString, ThingWithMap.class);
Assertions.assertThat(twnRx.getMyMap1().get("mk1")).isInstanceOf(MapValue1.class);
Assertions.assertThat(twnRx.getMyMap2().get("mk1")).isInstanceOf(LinkedHashMap.class);
```

This is exactly what I need, but I had difficulty because the TypeInfo implemetation govern deserialisation as well as serialisation
This means that if I attempt to derserialise `myMap1` without any embedded type info, it will fail.

```
com.fasterxml.jackson.databind.exc.InvalidTypeIdException: Could not resolve subtype of [simple type, class java.lang.Object]: missing type id property '_class' (for POJO property 'myMap1')
```

The difficulty I had was with global settings (such as the `TypeResolverBuilder`) then expects TypeInfo on ALL `Map<?,Object>` incoming json.   We have clients that don't add `"_class": "x.y.z.Class"` and it'd be a pain to weave this in across the board.

(We just accept the Map<String,Object> with the default Jackson behaviour of marshalling into a nested LinkedHashMap tree and marshall these outselves.)

Gladly now with your pointers I can selectively sprinkle this pattern through the code base as required.

P.S.  Yes - I appreciate that lombok has consequences - it's one of the first things to disable, in troubleshooting,

Thanks again.

**Reply by cowtowncoder** [reply]
@robshep Ah! I misread your comment -- I incorrectly thought it implied that combination wouldn't work. But reading it again I understood what you actually meant :)

Very glad you can make things work and that I was able to help. Best of luck!
